docs/Dialects/LoopSchedule/LoopSchedule.md

```diff
@@ -17,6 +17,88 @@ A primary goal of the `loopschedule` dialect, as opposed to many other High-Leve
+      %3 = arith.addi %arg1, %c1 : index
+      %4 = memref.load %arg0[%arg1] : memref<10xi32>
+      loopschedule.register %3, %4 : index, i32
+    } : index, i32
+    %2 = loopschedule.pipeline_stage start = 1 {
+      %3 = arith.addi %1#1, %arg2 : i32
+      pipeline.register %3 : i32
+    } : i32
+    loopschedule.terminator iter_args(%1#0, %2), results(%2) : (index, i32) -> i32
+  }
+  return %0 : i32
+}
+```
+
+A pipeline while op first defines the initial values for the iter_args. iter_args are values that will
+be passed back to the first stage after the last stage of the pipeline. The pipeline also defines a
+specific, static II. Each pipeline stage in the do block represents a series of ops run in parallel.
+Values are registered at the end of a stage and passed out as results for future pipeline stages to
+use. Each pipeline stage must have a defined start time, which is the number of cycles between the 
+start of the pipeline and when the first valid data with the available as input to that stage.
+Finally, the terminator is called with the iter_args for the next iteration and the result values
+that will be returned when the pipeline completes. Even though the terminator is located at the
+end of the loop body, its values are passed back to a previous stage whenever needed. We do not
+need to wait for an entire iteration to finish before iter_args become valid for the next iteration.
+
+Multi-cycle and pipelined ops can also be supported in pipeline while loops. In the following example,
+assume the multiply op is bound to a 3-stage pipelined multiplier:
+
+```
+func.func @test1(%arg0: memref<10xi32>, %arg1: memref<10xi32>) {
+  %c0 = arith.constant 0 : index
+  %c1 = arith.constant 1 : index
+  %c10 = arith.constant 10 : index
+  %c1_i32 = arith.constant 1 : i32
+  loopschedule.pipeline_while II = 1 iter_args(%arg2 = %c0) : (index, i32) -> () {
+    %1 = arith.cmpi ult, %arg1, %c10 : index
+    loopschedule.register %1 : i1
+  } do {
+    %1:2 = loopschedule.pipeline_stage start = 0 {
+      %3 = arith.addi %arg1, %c1 : index
+      %4 = memref.load %arg0[%arg2] : memref<10xi32>
+      loopschedule.register %3, %4 : index, i32
+    } : index, i32
+    %2:2 = loopschedule.pipeline_stage start = 1 {
+      %3 = arith.muli %1#1, %c1_i32 : i32
+      pipeline.register %3, %1#0 : i32
+    } : i32
+    loopschedule.pipeline_stage start = 4 {
+      memref.store %2#0, %arg0[%2#1] : i32
+      pipeline.register
+    } : i32
+    loopschedule.terminator iter_args(%1#0), results() : (index, i32) -> ()
+  }
+  return
+}
+```
+
+Here, the II is still 1 because new values can be introduced to the multiplier every cycle. The last
+stage is delayed by 3 cycles because of the 3 cycle latency of the multiplier. Because different ops
+can be mapped to many different physical implementations, the `pipeline_while` op is tightly coupled
+to the CIRCT scheduling and Calyx lowering infrastructures. This can be addressed in the future with
```

Comment:
I don't see any notion of tight coupling in the provided IR example and by extension the dialect itself. Isn't it rather that there - for now - only exists a _lowering_ to Calyx which understands the semantics that this operation tries to express (an `arith.muli` inside a `loopschedule.pipeline_stage` wherein its results are only used after 3 cycles)?

Reply:
I think there is some tight coupling here, but I also wouldn't point out Calyx necessarily. In the current scheme, the latencies of operators are not represented in the IR. That's the real problem, which has always made this IR coupled to the operator library used by the scheduling infrastructure. Right now, the only (implicit) operator library is Calyx, because that's the only lowering we have, and the scheduling problem is constructed assuming Calyx is the lowering.

We have paths out of this (put latencies in the IR, a-la filament, or finally define a proper operator library mechanism for CIRCT and use it). But I do think mentioning the coupling here is worthwhile.

Reply:
Yeah what Mike said here is what I was referring to. The lowering pass assumes an implicit Calyx operator library that is not encoded in the IR at the level of `loopschedule`. There's probably a better way to word this in the docs, I will look at how to improve this.